src/generate_features.py

```diff
@@ -1,3 +1,215 @@
 import logging
+import os
+from multiprocessing import Pool
+import sys
+import boto3
+import matplotlib.pyplot as plt
+import numpy as np
+import pandas as pd
+import yaml
+from sklearn import preprocessing
+from sklearn.cluster import KMeans
+from sklearn.metrics import calinski_harabaz_score as score_ch
+from sklearn.mixture import GaussianMixture as GMM
 
+
+logging.basicConfig(format="%(asctime)s - %(name)s - %(levelname)s - %(message)s")
 logger = logging.getLogger(__name__)
+logger.setLevel(logging.DEBUG)
+
+ROOT = os.getcwd()
+logger.debug('Project Root = %s', ROOT)
+
+with open(os.path.join(ROOT, 'config', 'features_config.yml'), 'r') as f:
+    config = yaml.load(f)
+with open(os.path.join(ROOT, 'data', 'external', 'cats.yml'), 'r') as f:
+    cat_map = yaml.load(f)
+    cats = list(cat_map.keys())
+logger.debug('%d categories loaded: %s', len(cats), cats)
+
+if config.get('read_from') == 'local':
+    data_dir = os.path.join(ROOT, 'data', 'external')
+elif config.get('read_from') == 's3':
+    data_dir = 's3://{}/data/external'.format(config.get('s3_bucket-name'))
+
+
+def load_data(data_dir, file_name):
+    path = os.path.join(data_dir, file_name)
+    logger.debug('Reading data from %s', path)
+    return pd.read_csv(path)
+
+
+def clean(X):
+    nans = np.isnan(X).any(axis=1)
+    logger.debug('Removing %d rows with nan values in order features (X)',
+                 sum(nans))
+    return X[~nans]
+
+
+def gmm_clust(X, n_components, cov_type):
+    logger.info('Clustering GMM with n_components=%d and cov_type=%s',
+                n_components, cov_type)
+    gmm = GMM(n_components=n_components, covariance_type=cov_type)
+    gmm.fit(X)
+    labels = gmm.predict(X)
+    return labels
+
+
+def km_clust(X, k):
+    logger.debug('Clustering KMeans with k=%d', k)
+    km = KMeans(X, k)
+    km.fit(X)
+    labels = km.labels_
+    return labels
+
+
+def gen_plots(order_types):
```

Comment:
where would saved images be/what purpose would they serve?

Reply:
the images will help to describe the order_types (heatmaps representing what features characterize a particular cluster). These may be used in the webapp directly, or at the least will help illuminate the analysis results